pkg/metadata/annotations.go

```diff
@@ -145,6 +145,14 @@ const (
 	// reconciler-manager will create the reconciler with the hydration-controller
 	// sidecar container.
 	RequiresRenderingAnnotationKey = configsync.ConfigSyncPrefix + "requires-rendering"
+
+	// RequiresNSControllerAnnotationKey is the annotation key set on R*Sync
+	// object to indicate whether the source of truth contains at least one
+	// NamespaceSelector using the dynamic mode, which requires the Namespace
+	// controller. The reconciler writes the value of this annotation and the
+	// reconciler-manager reads it. If set to true, the reconciler-manager will
+	// create the reconciler with the Namespace controller in the reconciler container.
+	RequiresNSControllerAnnotationKey = configsync.ConfigSyncPrefix + "requires-ns-controller"
```

Comment:
I just found that this page needs updating https://cloud.google.com/anthos-config-management/docs/reference/labels-and-annotations

Also, ns controller is an implementation detail. Are we planning to keep that architecture? If that will be gone with the controller-runtime refactor, this annotation should talk about the feature (user facing) instead of the controller (impl detail).

Reply:
Yes, the doc update for that page is on the list.

What about `requires-dynamic-ns-selector`?

Reply:
Minor tweak: `has-dynamic-ns-selector` (the RootSync has dynamic ns selector, instead of requires)

Reply:
`has-dynamic-ns-selector` indicates the current mode of NamespaceSelector in the RootSync's source repo, which may not match the actual mode.
What about `dynamic-ns-selector-enabled`, which implies the state of a former event. The annotation might be enabled by a previous state. If it doesn't match the current state, it should be updated.